ons-spark/pyspark-intro/f-col.md

```diff
@@ -0,0 +1,253 @@
+|129971-26092017|         Cat|
++---------------+------------+
+only showing top 5 rows
+```
+(f-col:example-2)=
+### Example 2: Filter on a new column
+
+Read in the animal rescue data:
+````{tabs}
+```{code-tab} py
+rescue = spark.read.parquet(rescue_path).select("incident_number", "animal_group")
+```
+````
+Let's create a new column, `animal_group_upper`, which consists of the `animal_group` in uppercase.
+
+If we try and immediately filter on this column using `rescue.animal_group_upper`, it won't work. This is because we have yet to define the column in `rescue`.
+````{tabs}
+```{code-tab} py
+try:
+    (rescue
+     .withColumn("animal_group_upper", F.upper(rescue.animal_group))
+     .filter(rescue.animal_group_upper == "CAT")
+     .show(5))
+except AttributeError as e:
+    print(e)
+```
+````
+
+```plaintext
+'DataFrame' object has no attribute 'animal_group_upper'
+```
+We could split this statement up over two different lines:
+````{tabs}
+```{code-tab} py
+rescue_upper = rescue.withColumn("animal_group_upper", F.upper(rescue.animal_group))
+rescue_upper.filter(rescue_upper.animal_group_upper == "CAT").show(5)
+```
+````
+
+```plaintext
++---------------+------------+------------------+
+|incident_number|animal_group|animal_group_upper|
++---------------+------------+------------------+
+|       80771131|         Cat|               CAT|
+|       43051141|         Cat|               CAT|
+|126246-03092018|         Cat|               CAT|
+|       17398141|         Cat|               CAT|
+|129971-26092017|         Cat|               CAT|
++---------------+------------+------------------+
+only showing top 5 rows
+```
+Using `F.col()` is instead is much neater:
+````{tabs}
+```{code-tab} py
+(rescue
+    .withColumn("animal_group_upper", F.upper(F.col("animal_group")))
+    .filter(F.col("animal_group_upper") == "CAT")
+    .show(5))
+```
+````
+
+```plaintext
++---------------+------------+------------------+
+|incident_number|animal_group|animal_group_upper|
++---------------+------------+------------------+
+|       80771131|         Cat|               CAT|
+|       43051141|         Cat|               CAT|
+|126246-03092018|         Cat|               CAT|
+|       17398141|         Cat|               CAT|
+|129971-26092017|         Cat|               CAT|
++---------------+------------+------------------+
+only showing top 5 rows
+```
+(f-col:example-3)=
+### Example 3: Ensuring you are using the latest values
+
+Using `df.column_name` can also result in bugs when you think you are referencing the latest values, but are actually using the original ones. Here, the values in `animal_group` are changed, but `rescue` is yet to be redefined, and so the old values are used. As such no data is returned:
+````{tabs}
+```{code-tab} py
+rescue = spark.read.parquet(rescue_path).select("incident_number", "animal_group")
+rescue.withColumn("animal_group", F.upper(rescue.animal_group)).filter(rescue.animal_group == "CAT").show(5)
```

Comment:
consider putting this on multiple lines. next code cell with similar code looks much nicer

Reply:
Done